### PR TITLE
[v2int/2.2] GC tombstone fixes

### PR DESCRIPTION
**Pull request by markfields**
These are cherry-picked commits from main that happened after the release branch forked, fixing some GC bugs we need fixed before partners can pick up Tombstone work - and they will improve existing telemetry signals too.

### CODE REVIEW

**Review by agarwal-navin — APPROVED** [review]

**Comment by markfields** [comment]
Waiting until Monday to merge to avoid interfering with more urgent release-blocking fixes coming in today.  There are 1-2 other changes needed to make good use of these fixes as well, so it makes sense to have a separate patch release for this GC work.

**Comment by markfields** [comment]
We ended up just waiting for internal.2.3

**Pull request closed without merging**
